### PR TITLE
Dealing with multiple atoms (coordinates/charges)

### PR DESCRIPTION
**Pull request by Rafael-G-C**
the code can now get the gradient and the hessian from a multi coordinate/charge input, such input is simplified as a numpy array for now ex.(coordinates = np.array([[1.0,2.0,3.0],[2.0,1.0,3.0],[3.0,1.0,2.0],[5.0,4.0,3.0],[4.0,3.0,5.0]]) charges = [1.0,1.0,1.0,1.0,1.0]

To keep it simple I left the function E_calc (on jax_test3.py) as is and changed get_grad and get_hessian to deal with the multiple coordinates.

After the atom check it differentiates the E_calc function, feeds it the coordinates and repeats.

analytical.py was also changed to deal with the new input

### CODE REVIEW

**Comment by robertodr** [comment]
Your local changes are conflicting with the remote ones. You need to resolve them and you can do so in a web editor by clicking the "Resolve conflicts" button. In the editor you will see both the local and remote versions and you can decide which one to keep (I guess your local version).

**Comment by Rafael-G-C** [comment]
done

File: jax_test3.py [hunk 1]
```diff
@@ -8,34 +8,50 @@
 
 def E_calc(x_a, y_a, z_a, x_b, y_b, z_b, c_a, c_b):
     d = jnp.sqrt((x_a - x_b) ** 2 + (y_a - y_b) ** 2 + (z_a - z_b) ** 2)
-    E = (c_a * c_b) / d
-    return E
+    energy = (c_a * c_b) / d
+    return energy
```
Comment (robertodr) [inline jax_test3.py]: This needs to be implemented to accept a matrix of coordinates and a vector of charges. We're not calculating the correct energy/gradient/Hessian otherwise.
The inner loop could look something like:
```python
energy = 0.0
natoms = charges.size
for A in range(natoms):
    for B in range(A):
         d = jnp.linalg.norm(coordinates[A, :] - coordinates[B, :])
         energy += charges[A] * charges[B] / d
```

File: tests/analytical.py [hunk 2]
```diff
@@ -1,88 +1,113 @@
 import numpy as np
 
 
-def gradient(x_a, y_a, z_a, x_b, y_b, z_b, c_a, c_b):
+def gradient(coordinates,charges):
     """Computes the gradient."""
+    number_coordinates = len(coordinates)
+    eval_E = []
```
Comment (robertodr) [inline tests/analytical.py]: We know already the size of the gradient vector: 3 times the number of atoms (the x, y, z derivatives for each atom) We can use a numpy array directly, rather than first a list and then a `np.array`:
```python
eval_E = np.zeros_like(coordinates)
```
`zeros_like` creates a numpy array of zeros with the same shape (number of rows and columns) of the parameter. In this case, the gradient will be a matrix like the coordinates matrix.
Documentation for `zeros_like` here: https://numpy.org/doc/stable/reference/generated/numpy.zeros_like.html

File: tests/analytical.py [hunk 3]
```diff
@@ -1,88 +1,113 @@
...
+    for atom_a in range(number_coordinates):
+        for atom_b in range(number_coordinates):
+            if atom_a < atom_b:
+                cxc = charges[atom_a] * charges[atom_b]
+                xyz_squared = ((coordinates[atom_a,0] - coordinates[atom_b,0]) ** 2 + (coordinates[atom_a,1] - coordinates[atom_b,1]) ** 2 + (coordinates[atom_a,2] - coordinates[atom_b,2]) ** 2) ** (3 / 2)
```
Comment (robertodr) [inline tests/analytical.py]: You can get this a bit more easily and concisely with:
```python
xyz_32 = np.linalg.norm(coordinates[atom_a, :] - coordinates[atom_b, :])**3
```
Step by step:
1. `coordinates[atom_a, :] - coordinates[atom_b, :]` takes the difference between the rows with indices `atom_a` and `atom_b`. Numpy indexing notation documentation: https://numpy.org/doc/stable/user/basics.indexing.html
2. The `norm` function calculates the norm of a vector.
3. We take the third power.

File: tests/analytical.py [hunk 4]
```diff
@@ -1,88 +1,113 @@
...
 
-    cxc = c_a * c_b
-    xyz_squared = ((x_a - x_b) ** 2 + (y_a - y_b) ** 2 + (z_a - z_b) ** 2) ** (3 / 2)
+                d_ax = -(coordinates[atom_a,0] - coordinates[atom_b,0]) * cxc / xyz_squared
+                eval_E.append(d_ax)
```
Comment (robertodr) [inline tests/analytical.py]: Once `eval_E` becomes a proper numpy array, you can set its elements by indexing:
```python
eval_E[atom_a, 0] = d_ax
eval_E[atom_a, 1] = d_ay
eval_E[atom_a, 2] = d_az 
```
you can even remove the intermdiate variables `d_ax`, `d_ay`, `d_az` and assign directly to `eval_E`. If you use the `:` then it can be done at once:
```python
eval_E[atom_a, :] = -(coordinates[atom_a, :] - coordinates[atom_b, :]) * cxc / xyz_32
```

File: tests/analytical.py [hunk 5]
```diff
@@ -1,88 +1,113 @@
 import numpy as np
 
 
-def gradient(x_a, y_a, z_a, x_b, y_b, z_b, c_a, c_b):
+def gradient(coordinates,charges):
```
Comment (robertodr) [inline tests/analytical.py]: I don't think the expression for the gradient you are implementing is entirely correct. If you only have two atoms, then:
dffa0427077639c231eaefce54e63473
and the gradient with respect to atom 0 (for example) is:
89f21e761da3a8768b7a67ebfc519333

Now, if we have 3 atoms:
9cc49e84e878a8f3e26ef42a0db5ea47
and the gradient with respect to atom 0:
04bded84ac9d28abd07d804de4776260

So this line:
```python
d_ax = -(coordinates[atom_a,0] - coordinates[atom_b,0]) * cxc / xyz_squared
```
should really be:
```python
d_ax -= (coordinates[atom_a,0] - coordinates[atom_b,0]) * cxc / xyz_squared
```
because you're accumulating the result of differentiating with respect to all atoms other than atom 0.

**Comment by robertodr** [comment]
I don't think this implementation is entirely correct at the moment. Try to:
1. Implement the energy evaluation function such that it accepts a matrix of coordinates and an array of charges.
2. Derive the formula for the gradient when you have N atoms. This should help fix the analytical function.

**Comment by Rafael-G-C** [comment]
great! will start making the changes

**Review by robertodr — APPROVED** [review]
Excellent!
